blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/listeners/DownstreamJobListener.java

```diff
@@ -0,0 +1,34 @@
+package io.jenkins.blueocean.listeners;
+
+import hudson.Extension;
+import hudson.model.Cause;
+import hudson.model.CauseAction;
+import hudson.model.Run;
+import hudson.model.TaskListener;
+import hudson.model.listeners.RunListener;
+import io.jenkins.blueocean.service.embedded.DownstreamJobAction;
+
+/**
+ * Listens to creation of jobs that are triggered by an upstream job, much like BuildTriggerListener from
+ * pipeline-build-step. Unlike BuildTriggerListener, here we just add an action to the upstream job for later retrieval
+ * rather than injecting information into its build log, in order to be a little more flexible at the cost of a bit more
+ * work later to retrieve the details.
+ */
+@Extension
+public class DownstreamJobListener extends RunListener<Run<?, ?>> {
+
+    @Override
+    public void onStarted(Run<?, ?> run, TaskListener listener) {
+        for (CauseAction action : run.getActions(CauseAction.class)) {
+            for (Cause cause : action.getCauses()) {
+                if (cause instanceof Cause.UpstreamCause) {
+                    Run triggerRun = ((Cause.UpstreamCause) cause).getUpstreamRun();
+                    if (triggerRun != null) {
+                        // Add a link from the upstream job to this newly spawned one, to be retrieved later
+                        triggerRun.addAction(new DownstreamJobAction(run));
```

Comment:
So this is going to modify the upstream `build.xml`.

Not a problem exactly, but not what we would usually do to record Pipeline build metadata. The better approach would be for `pipeline-build-step` to record a `[FlowNode]Action` on the node corresponding to the `build` step, so that you have a record of exactly when this downstream build was started. (Currently it adds only a `LabelAction`.) Then you could retrieve a flat list of all downstream builds of a given upstream build, as in this PR, _or_ more detailed information about (say) downstream builds in a given branch.

Reply:
I will look into this, thanks.

Reply:
I can get on that tomorrow, if you’d like.

Reply:
Worth mentioning, though, that there is always a bit of a challenge in associating the upstream `build` with the actual downstream run, which may not start for a while (freestyle waiting on an agent, a project not allowing concurrent builds, etc...), which is part of why we’ve historically tracked the upstream/downstream relationship from the downstream run/job rather than from the upstream.

Reply:
@abayer that's why I took a similar approach to the downstream-build listener in `pipeline-build-step`. I think we want to continue doing it this way, but look at attaching the data to the flownode rather than the run itself, as Jesse suggested. Not really sure how I'll go about that yet, but that's my current thinking.